%\VignetteIndexEntry{INSPEcT}
%\VignettePackage{INSPEcT}
%\VignetteDepends{INSPEcT}
%\VignetteEngine{knitr::knitr}
\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}

<<style-knitr, eval=TRUE, echo=FALSE, results="asis">>=

BiocStyle::latex()
@

\newcommand{\Rmethod}[1]{{\Rfunction{#1}}}

\title{INSPEcT - INference of Synthesis, Processing and dEgradation rates from Transcriptome data}
\author{de Pretis S. - Furlan M. - Pelizzola M.}

\begin{document}

\maketitle

\tableofcontents

\section{Introduction}

  \Rpackage{INSPEcT} is an R/Bioconductor compliant solution for the study of RNA transcriptional dynamics
  from RNA-seq data. The package supports the analysis of several experimental designs. It can handle
  both time-course data or the comparison of steady-state conditions, and the presence or absence of the nascent RNA 
  library. \\
  According to the experimental design the software returns different outputs. A schema of the possible configurations and 
  related outputs is reported in Table \ref{table:INSPEcTanalysisDesigns}. 
  This scheme is supposed to guide the reader across the vignette making the individuation of the desired examples easier, 
  however, the authors recommend at least one complete reading of the document to understand: 
  the references between sections, the logic behind \Rpackage{INSPEcT} and the potential of the package. 
  For more details about \Rpackage{INSPEcT}, examples of its application and extended methods description 
  refer to REFERENCE TO THE PAPER.

  \begin{table}\label{table:INSPEcTanalysisDesigns}
  \caption{Table reporting the experimental designs suitable for INSPEcT}

  \begin{tabular}{cl|c|c|c|}
  \cline{3-5}
                                              &  & \multicolumn{3}{c|}{\textbf{Total and nascent RNA}}              \\ \cline{3-5}
                                              &  & \textit{Synthesis} & \textit{Processing} & \textit{Degradation}  \\ \cline{1-1} \cline{3-5}
  \multicolumn{1}{|c|}{\textbf{Steady state}} &  & Yes                & Yes                 & Yes                   \\ \cline{1-1} \cline{3-5}
  \multicolumn{1}{|c|}{\textbf{Time-course}}  &  & Yes                & Yes                 & Yes                   \\ \cline{1-1} \cline{3-5}
  \end{tabular}

  \begin{tabular}{cl|c|c|c|}
  \cline{3-5} 
                                              &  & \multicolumn{3}{c|}{\textbf{Total RNA}}                         \\ \cline{3-5}
                                              &  & \textit{Synthesis} & \textit{Processing} & \textit{Degradation} \\ \cline{1-1} \cline{3-5}
  \multicolumn{1}{|c|}{\textbf{Steady state}} &  & No                 & \multicolumn{2}{c|}{Ratio $\frac{Processing}{Degradation}$}                 \\ \cline{1-1} \cline{3-5}
  \multicolumn{1}{|c|}{\textbf{Time-course}}  &  & Yes                & Yes                 & Yes                  \\ \cline{1-1} \cline{3-5}
  \end{tabular}

  \end{table}

\section{Mature and Nascent time-course RNA-seq data}

  \subsection{Transcripts expression quantification}
    The \Rpackage{INSPEcT} analysis starts with the quantification of exons and introns expressions and 
    the estimation of the variance of the quantification from the replicates, which are mandatory.\\

    <<INSPEcT_loading,eval=TRUE,echo=TRUE,message=FALSE,warnings=FALSE>>=
    library(INSPEcT)
    @

    In particular, INSPEcT quantification of transcripts expression has different entry points according to the 
    source of data available:
    \begin{itemize}
      \item BAM or SAM files (provided with an annotation object of class TxDb)

          <<quantifyExpressionsFromBAMs,eval=FALSE,message=FALSE,warnings=FALSE>>=
        
          require(TxDb.Mmusculus.UCSC.mm9.knownGene)
          txdb <- TxDb.Mmusculus.UCSC.mm9.knownGene

          bamfiles_paths <- system.file('extdata/',
            c('bamRep1.bam','bamRep2.bam','bamRep3.bam','bamRep4.bam'), 
            package='INSPEcT')
          
          exprFromBAM <- quantifyExpressionsFromBAMs(
                     txdb=txdb
                     ,BAMfiles=bamfiles_paths
                     ,DESeq2=TRUE
                     ,experimentalDesign=c(0,0,1,1))
          
          @

      \item exonic and intronic read counts

          <<quantifyExpressionsFromCountsNascentTimeCourse,eval=TRUE,message=FALSE,warnings=FALSE>>=
          
              data('allcounts', package='INSPEcT')
              data('featureWidths', package='INSPEcT')
              data('libsizes', package='INSPEcT')

              nascentCounts<-allcounts$nascent
              matureCounts<-allcounts$mature

              expDes<-rep(c(0,1/6,1/3,1/2,1,1.5,2,4,8,12,16),3)

              nasExp_DESeq2<-quantifyExpressionsFromTrCounts(
                      allcounts=nascentCounts
                      ,libsize=nascentLS
                      ,exonsWidths=exWdths
                      ,intronsWidths=intWdths
                      ,experimentalDesign=expDes)

              matExp_DESeq2<-quantifyExpressionsFromTrCounts(
                      allcounts=matureCounts
                      ,libsize=totalLS
                      ,exonsWidths=exWdths
                      ,intronsWidths=intWdths
                      ,experimentalDesign=expDes)

          @

      \item exonic and intronic transcripts abundances
    \end{itemize}
    It can be done starting from BAM or SAM files through the method \Rmethod{quantifyExpressionsFromBAMs} which exploits the package \Rpackage{GenomicAlignments} to convert RNA-seq reads in counts and then the method \Rmethod{quantifyExpressionsFromTrCounts} to make them RPKMs. The evaluation of the variance is part of this initial stage of the analysis too and it can be done using \Rpackage{DESeq2} (\textit{DESeq2 = TRUE}) or \Rpackage{plgem} (\textit{DESeq2 = FALSE}); the second option requires to select a condition with replicates from the experimental design to use for the variance sampling (\textit{varSamplingCondition}). Two expression matrices, together with the associated variance matrices, constitute the output of the method.\\
    The user can also access the pipeline providing counts data, calling directly \Rmethod{quantifyExpressionsFromTrCounts}, or expression data (like RPKMs) through \Rmethod{quantifyExpressionsFromTrAbundance}; in the latter case \Rpackage{plgem} is the only choice for variance evaluation.\\
    In this configuration we assume the coexistence of two populations of transcripts: nascent ($L$) and mature ($R$) RNA molecules. The first group includes all the molecules synthesized during a specific span of time (labelling time $t_L$), which ends with the preparation of the sample for the sequencing, while the second one represents the RNA present in the biological system before $t_L$. Regardless of the role of $L$ and $R$ in the \Rpackage{INSPEcT} analysis framework, which will be clarified soon, the existence of two distinct populations of RNAs implies that the aforementioned procedure must be repeated twice.




Enjoy!

\section{About this document}

<<r>>=

print("r")

sessionInfo()
@

\end{document}