%\VignetteIndexEntry{INSPEcT}
%\VignettePackage{INSPEcT}
%\VignetteDepends{INSPEcT}
%\VignetteEngine{knitr::knitr}
\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}

\RequirePackage[]{/home/mattia/R/x86_64-pc-linux-gnu-library/3.4/BiocStyle/resources/tex/Bioconductor2}
\AtBeginDocument{\bibliographystyle{/home/mattia/R/x86_64-pc-linux-gnu-library/3.4/BiocStyle/resources/tex/unsrturl}}


\newcommand{\Rmethod}[1]{{\Rfunction{#1}}}

\title{INSPEcT - INference of Synthesis, Processing and dEgradation rates in Time-course analysis}
\author{AUTORI}

\begin{document}

\maketitle

\tableofcontents

%
\section{Introduction}

\Rpackage{INSPEcT} provides an R/Bioconductor compliant solution for the study of dynamic transcriptional regulatory processes. Based on RNA-seq data, \Rpackage{INSPEcT} determines mRNA synthesis, degradation and pre-mRNA processing rates over time for each gene, genome-wide. Moreover, the \Rpackage{INSPEcT} modeling framework allows the identification of gene-level transcriptional regulatory mechanisms, determining which combination of sythesis, degradation and processing rates is most likely responsible for the observed mRNA level over time.
\Rpackage{INSPEcT} provides two different analysis approaches: the first one requires total RNA-seq data (No4su configuration) while the second one (4su configuration) also needs 4su-seq data.
Regarding the second scenario, 4su-seq is a recent experimental technique developed to measure the concentration of nascent mRNA and for the genome-wide inference of gene-level synthesis rates. During a short pulse (typically few minutes), cells medium is complemented with 4-Thiouridine (4su), a naturally occurring modified uridine that is incorporated within growing mRNA chains with minimal impact on cell viability. The chains which have incorporated the uridine variant (the newly synthesized ones) can be isolated from the total RNA population by biotinylation and purification with streptavidin-coated magnetic beads, followed by sequencing. In both cases, the main set of steps in the \Rpackage{INSPEcT} workflow is as follows:
\begin{itemize}
	\item Exonic and intronic RNA-seq RPKMs are determined for each gene and allow to quantify total mRNA and pre-mRNA (\Rfunction{makeRPKMsFromBams}, \Rfunction{makeRPKMsFromCounts}),
	\item synthesis, processing and degradation rates are obtained (\Rmethod{newINSPEcT}),
	\item rates, total mRNA and pre-mRNA concentrations are modeled for each gene to assess which of the rates, if any, determined changes in mRNA levels (\Rmethod{modelRates}).
\end{itemize}
For the 4su based analysis, simulated data that recapitulate rate distributions, their variation over time and their pair-wise correlations can be created and used to evaluate the performance of the method (\Rmethod{makeSimModel}, \Rmethod{makeSimDataset}, \Rmethod{rocCurve}). The artificial data can also be exploited to evaluate the performances of the no-4su configuration, however, they can not be generated without the 4su-seq data.\\
Whitin this vignette, some sections are explicitly divided in two parts, one for each aforementioned analysis configuration, in order to seep up the consultation. However, the authors recomend at least one complete reading of the document to understand the logic behind \Rpackage{INSPEcT} and the potential of the package. For details regarding \Rpackage{INSPEcT} extended methods description, refer to .... ... .
%
\section{Quantification of Exon and Intron features}

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.941, 0.941, 0.941}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{library}\hlstd{(INSPEcT)}
\end{alltt}
\end{kframe}
\end{knitrout}

The \Rpackage{INSPEcT} framework includes function to quantify exon and intron features. The function \Rmethod{makeRPKMsFromBams} builds an annotation for exons and introns, either at the level of transcripts or genes, and count reads falling on the two different annotations in each provided file in BAM or SAM format. This function prioritize the exon annotation, meaning that reads which fall on exon are not counted for introns, in case of overlap. Canonical RPKMs are then evaluated and returned as output together with read counts and annotations. \Rpackage{INSPEcT} also allows to recompute RPKMs and counts through \Rpackage{DESeq2}, the experimental design is provided by the argument \textit{ temporalDesign }. If the \Rpackage{DESeq2} analysis is performed the package will be exploited also to compute the variances of the experimental data in the next steps of the pipeline; for this porpuse the output of the function contains the additional list of parameters \textit{ dispersion\_parameters\_DESeq2 }.

\subsubsection{4su working example}



































































